pennylane_lightning_gpu/src/simulator/StateVectorCudaMPI.hpp

```diff
@@ -941,6 +960,379 @@ class StateVectorCudaMPI
+     * matrix. Array of size numNNZ
+     * @param values_ptr Pointer to the array of the non-zero elements
+     *
+     * @return auto A vector of vector of CSRMatrix.
+     */
+    template <class index_type>
+    auto splitCSRMatrix(const size_t &num_row_blocks,
+                        const size_t &num_col_blocks,
+                        const index_type *csrOffsets_ptr,
+                        const index_type *columns_ptr,
+                        const std::complex<Precision> *values_ptr)
+        -> std::vector<std::vector<CSRMatrix<Precision, index_type>>> {
+
+        std::vector<std::vector<CSRMatrix<Precision, index_type>>>
+            splitSparseMatrix(
+                num_row_blocks,
+                std::vector<CSRMatrix<Precision, index_type>>(num_col_blocks));
+
+        size_t num_rows = size_t{1} << this->getTotalNumQubits();
+        size_t row_block_size = size_t{1} << this->getNumLocalQubits();
+        size_t col_block_size = row_block_size;
+
+        // Add OpenMP support here later.
+        for (size_t row = 0; row < num_rows; row++) {
+            for (size_t col_idx = static_cast<size_t>(csrOffsets_ptr[row]);
+                 col_idx < static_cast<size_t>(csrOffsets_ptr[row + 1]);
+                 col_idx++) {
+
+                size_t current_global_row = row;
+                size_t current_global_col = columns_ptr[col_idx];
+                std::complex<Precision> current_val = values_ptr[col_idx];
+
+                size_t block_row_id = current_global_row / row_block_size;
+                size_t block_col_id = current_global_col / col_block_size;
+
+                size_t local_row_id = current_global_row % row_block_size;
+                size_t local_col_id = current_global_col % col_block_size;
+
+                if (splitSparseMatrix[block_row_id][block_col_id]
+                        .getCsrOffsets()
+                        .size() == 0) {
+                    splitSparseMatrix[block_row_id][block_col_id]
+                        .getCsrOffsets() =
+                        std::vector<index_type>(row_block_size + 1, 0);
+                }
+
+                splitSparseMatrix[block_row_id][block_col_id]
+                    .getCsrOffsets()[local_row_id + 1]++;
+                splitSparseMatrix[block_row_id][block_col_id]
+                    .getColumns()
+                    .push_back(local_col_id);
+                splitSparseMatrix[block_row_id][block_col_id]
+                    .getValues()
+                    .push_back(current_val);
+            }
+        }
+
+        // Add OpenMP support here later.
+        for (size_t block_row_id = 0; block_row_id < num_row_blocks;
+             block_row_id++) {
+            for (size_t block_col_id = 0; block_col_id < num_col_blocks;
+                 block_col_id++) {
+                for (size_t i0 = 1;
+                     i0 < splitSparseMatrix[block_row_id][block_col_id]
```

Comment:
Since `splitSparseMatrix[block_row_id]` is not-changing within the second & third loops, we can likely define an intermediate reference variable to it outside the block_col_id loop, and use it throughout. It'll prevent a bunch of look-ups. The same with the `[block_col_id]` look-up for the third loop.

Reply:
Done!